types/directive.go

```diff
@@ -24,6 +28,11 @@ type DirectiveDefinition struct {
 
 type DirectiveList []*Directive
 
+type DirectiveVisitor interface {
+	Before(ctx context.Context, directive *Directive, input interface{}) error
+	After(ctx context.Context, directive *Directive, output interface{}) (interface{}, error)
+}
```

Comment:
Do you think we could use generics here to make this type-safe and avoid interface casting throughout the code? 🤔 
Something like:
```suggestion
type DirectiveVisitor[I any, O any] interface {
	Before(ctx context.Context, directive *Directive, input I) error
	After(ctx context.Context, directive *Directive, output O) (O, error)
}
```

Reply:
I know that this would require changes in this line:
```go
	visitors                 map[string]types.DirectiveVisitor
```
and we wouldn't be able to easily loop through all visitors, however, it would make the life of our users much better. All visitors behave the same and they differ just by the input and output types. Therefore, it'd be a nice candidate for generics. What do you think?

Reply:
Sure. I will try this and see how it feels.

Reply:
Yes, it is high time we updated that.

Reply:
I pushed up a commit (`a7c50d4`) that tries a pretty naive approach with generics. It doesn't compile. It seems like when combining the DirectiveVisitor interface with generics, we don't gain much in the way of type safety. The interface rules in Go mean that the particular implementations must match the type signature of the interface which leads to compile errors like this.

```./graphql_test.go:290:25: cannot use &customDirectiveVisitor{} (value of type *customDirectiveVisitor) as type types.DirectiveVisitor[any, any] in map literal:
	*customDirectiveVisitor does not implement types.DirectiveVisitor[any, any] (wrong type for After method)
		have After(ctx context.Context, directive *types.Directive, output string) (string, error)
		want After(ctx context.Context, directive *types.Directive, output any) (any, error)
```

Reply:
[1506179](https://github.com/graph-gophers/graphql-go/pull/543/commits/15061797b04c9e4a89b1427e81a26cd896df45da)

This latest commit adds a new file `types/directives.go` which will allow you to play around with the type system here. I don't have an intuition for Go's generics yet, so having this playground was helpful for me. I'm not sure there is a workable path to compile-time type safety because Go decided not to implement type covariance in its generics.

To be clear, the thing that's missing here is the ability to collect/abstract all Directives in a slice because we can't refer to them without specifying a type parameter. I'm going to keep working on this, just wanted to share my progress.

Reply:
I'm going to sleep since it's late here. I'll think some more on this tomorrow. I was thinking that we can still use `interface{}` inside of the library to combine all the visitors in a single slice, however, this will be behind the scenes. The users of the library will only use concrete types for input and output, right? I haven't tested if this would work, though. I might check tomorrow. In any case your initial PR was fine. I'm trying to think if we can improve it before we make it available for everyone to use.

Reply:
I did enough goofing around with generics to convince myself that they won't help us here. For posterity, here is a playground that demonstrates the two issues I see: https://go.dev/play/p/kIQdCQM1_cz

1) We cannot form a collection of concrete visitors with varied type signatures, there is no way to express their type. Go really wants us to specify a concrete type (instantiation). `interface{}` doesn't help us here, because the generic type checker expects the concrete values to have `type interface{}`
2) It's not clear how the library code (ie. exec.go) would discover what type to use when it calls the hooks. Since the hooks have concrete types, we must pass them values that type-check. We could do some risky run-time reflecty stuff here but that doesn't seem better than what we have now.